Detectors/CTP/reconstruction/include/CTPReconstruction/CTFCoder.h

```diff
@@ -151,27 +161,81 @@ o2::ctf::CTFIOSize CTFCoder::decode(const CTF::base& ec, VTRG& data, LumiInfo& l
+      }
+      if (CTPInputMask) {
+        if (digitsMap.count(irs)) {
+          if (digitsMap[irs].isInputEmpty()) {
+            digitsMap[irs].CTPInputMask = CTPInputMask;
+            // LOG(info) << "IR1:";
+            // digitsMap[irs].printStream(std::cout);
+          } else {
+            LOG(error) << "CTPInpurMask already exist:" << irs << " dig.CTPInputMask:" << digitsMap[irs].CTPInputMask << " CTPInputMask:" << CTPInputMask;
+          }
+        } else {
+          CTPDigit dig = {irs, CTPInputMask, 0};
+          digitsMap[irs] = dig;
+          // LOG(info) << "IR2:";
+          // digitsMap[irs].printStream(std::cout);
+        }
+      }
     } else { // correction would make IR prior to mFirstTFOrbit, skip
       itInp += CTFHelper::CTPInpNBytes;
+    }
+    if (checkIROK || canApplyBCShift(ir)) { // correction will be ok
+      auto irs = ir - mBCShift;
+      uint64_t CTPClassMask = 0;
+      for (int i = 0; i < CTFHelper::CTPClsNBytes; i++) {
+        CTPClassMask |= static_cast<uint64_t>(*itCls++) << (8 * i);
+      }
+      if (CTPClassMask) {
+        if (digitsMap.count(irs)) {
+          if (digitsMap[irs].isClassEmpty()) {
+            digitsMap[irs].CTPClassMask = CTPClassMask;
+            // LOG(info) << "TCM1:";
+            // digitsMap[irs].printStream(std::cout);
+          } else {
+            LOG(error) << "CTPClassMask already exist:" << irs << " dig.CTPClassMask:" << digitsMap[irs].CTPClassMask << " CTPClassMask:" << CTPClassMask;
+          }
+        } else {
+          CTPDigit dig = {irs, 0, CTPClassMask};
+          digitsMap[irs] = dig;
+          // LOG(info) << "TCM2:";
+          // digitsMap[irs].printStream(std::cout);
+        }
+      }
+    } else { // correction would make IR prior to mFirstTFOrbit, skip
       itCls += CTFHelper::CTPClsNBytes;
-      continue;
     }
-    auto& dig = data.emplace_back();
-    dig.intRecord = ir - mBCShift;
-    for (int i = 0; i < CTFHelper::CTPInpNBytes; i++) {
-      dig.CTPInputMask |= static_cast<uint64_t>(*itInp++) << (8 * i);
+  }
+  if (mDecodeInps) {
+    o2::pmr::vector<CTPDigit> digits;
+    o2::ctp::RawDataDecoder::shiftInputs(digitsMap, digits, mFirstTFOrbit);
+    for (auto const& dig : digits) {
+      data.emplace_back(dig);
     }
-    for (int i = 0; i < CTFHelper::CTPClsNBytes; i++) {
-      dig.CTPClassMask |= static_cast<uint64_t>(*itCls++) << (8 * i);
+  } else {
+    for (auto const& dig : digitsMap) {
+      data.emplace_back(dig.second);
     }
   }
   assert(itInp == bytesInput.end());
   assert(itCls == bytesClass.end());
   iosize.rawIn = header.nTriggers * sizeof(CTPDigit);
   return iosize;
 }
+///________________________________
+template <typename CTF = o2::ctp::CTF>
+bool CTFCoder::finaliseCCDB(o2::framework::ConcreteDataMatcher& matcher, void* obj)
+{
+  auto match = o2::ctf::CTFCoderBase::finaliseCCDB<CTF>(matcher, obj);
+  mBCShiftInputs = -o2::ctp::TriggerOffsetsParam::Instance().globalInputsShift;
+  return match;
+}
```

Comment:
Actually, instead of redefining this method, please redefine the 
```
void CTFCoder::updateTimeDependentParams(o2::framework::ProcessingContext& pc, bool askTree)
{
   o2::ctf::CTFCoderBase::updateTimeDependentParams(pc, askTree);
   mBCShiftInputs = -o2::ctp::TriggerOffsetsParam::Instance().globalInputsShift;
}
```

Reply:
updateTimeDependentParam is called for every TF. Is it waht we want ?

Reply:
Hm, this costs nothing, but ok, let's leave as it is. The only problem with the current code is that mBCShiftInputs will be set as many times as objects we are loading. But, since eventually it will be set correctly, it is not a problem.

Reply:
and this ?
template <typename CTF = o2::ctp::CTF>
bool CTFCoder::finaliseCCDB(o2::framework::ConcreteDataMatcher& matcher, void* obj)
{
  bool match = false;
  match = o2::ctf::CTFCoderBase::finaliseCCDB<CTF>(matcher, obj);
  if ((match = (matcher == o2::framework::ConcreteDataMatcher("CTP", "Trig_Offset", 0)))) {
    mBCShiftInputs = -o2::ctp::TriggerOffsetsParam::Instance().globalInputsShift;
    LOG(info) << "CTP BC shifts inputs:" << mBCShiftInputs << " TClasses:" << mBCShift;
  }
  return match;
}

Reply:
formally this is correct, but the ConcreteDataMatcher check is much more expensive than an extra call to o2::ctp::TriggerOffsetsParam::Instance().globalInputsShift, so leave it as it is now.

Reply:
ok, I will.